### PR TITLE
Generating .pyi file for IterDataPipes within the module

### PR DESCRIPTION
**Pull request by NivekT**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #224

This attempts to allow method hints to be shown to users when TorchData DataPipe is imported and used. See the comment below for one issue. It may be the best we can do for now.

If this change can lead to confusion, maybe we should land this PR instead: https://github.com/pytorch/pytorch/pull/70012
That PR adds all method hints to PyTorch Core, but the downside is that users will be suggested to use methods that are only available with TorchData installed.

In the long term, it will likely be better for all DataPipes to be in one single library.

Note: the Mypy error will disappear when this PR lands: https://github.com/pytorch/pytorch/pull/72829

Differential Revision: [D34243814](https://our.internmc.facebook.com/intern/diff/D34243814)

### CODE REVIEW

File: torchdata/datapipes/iter/__init__.pyi [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+from .util.extractor import CompressionType
+
+try:
+    import torcharrow
+except ImportError:
+    torcharrow = None
+
+T_co = TypeVar("T_co", covariant=True)
+
+class IterDataPipe(IterableDataset[T_co], metaclass=_DataPipeMeta):
```
Comment (NivekT) [inline torchdata/datapipes/iter/__init__.pyi:169]: This file is the generated .pyi file. There is one issue - the method hints for TorchData functional DataPipes only seem to show up when a TorchData-defined DataPipe is used.

For example, there is only PyTorch Core functional method hint here:
```python
dp = IterableWrapper([1,2,3])  # Only PyTorch Core method hints are shown
dp = dp.map(lambda x: x)  # Only PyTorch Core method hints are shown
dp = dp.filter(lambda _: False)  # Only PyTorch Core method hints are shown
```
There are both PyTorch Core and TorchData functional method hints after `Enumerator` is called:
```python
dp = IterableWrapper([1,2,3])  # Only PyTorch Core method hints are shown
dp = Enumerator(dp)  # From this point on, all method hints are visible
dp = dp.map(lambda x: x)  # All method hints are shown
dp = dp.filter(lambda _: False). # All method hints are shown
```

**Comment by ejguan** [comment]
I believe the ultimate solution would be adding `pyi.in` file in PyTorch Core that contains all functional API in PyTorch Core. And, when uses `import torch`, a `pyi` file is copied from `pyi.in`. Then, whenever users `import torchdata`, new functional API needs to be appended.

> In the long term, it will likely be better for all DataPipes to be in one single library.

I don't know if it's feasible as we want to keep BC.

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34243814).

**Comment by NivekT** [comment]
> I believe the ultimate solution would be adding `pyi.in` file in PyTorch Core that contains all functional API in PyTorch Core. And, when uses `import torch`, a `pyi` file is copied from `pyi.in`. Then, whenever users `import torchdata`, new functional API needs to be appended.
> 
> > In the long term, it will likely be better for all DataPipes to be in one single library.
> 
> I don't know if it's feasible as we want to keep BC.

If I understand correctly, you are suggesting generating the `.pyi` file at import time. Are there existing examples of this?


Aside from that, any thoughts on this PR as it is?

File: torchdata/datapipes/iter/__init__.pyi.in [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        cls, function_name: Any, cls_to_register: Any, enable_df_api_tracing: bool = ...
+    ): ...
+    def __getstate__(self): ...
+    def __reduce_ex__(self, *args: Any, **kwargs: Any): ...
+    @classmethod
+    def set_getstate_hook(cls, hook_fn: Any) -> None: ...
+    @classmethod
+    def set_reduce_ex_hook(cls, hook_fn: Any) -> None: ...
+    # Functional form of 'BatcherIterDataPipe'
+    def batch(self, batch_size: int, drop_last: bool = False, wrapper_class=DataChunk) -> IterDataPipe: ...
```
Comment (ejguan) [inline torchdata/datapipes/iter/__init__.pyi.in:42]: Why do we need to specify all functional APIs from PyTorch Core here?
Why not to use `gen_pyi` to generate these APIs as torchdata depends on torch?

Comment (NivekT) [reply torchdata/datapipes/iter/__init__.pyi.in]: Good catch - I have updated both PRs to make it possible to do that.

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34243814).

**Pull request closed without merging**